source/server/admin/prometheus_stats.cc

```diff
@@ -228,11 +218,29 @@ uint64_t PrometheusStatsFormatter::statsAsPrometheus(
 }
 
 bool PrometheusStatsFormatter::registerPrometheusNamespace(absl::string_view prometheus_namespace) {
-  if (std::regex_match(prometheus_namespace.begin(), prometheus_namespace.end(),
-                       namespaceRegex())) {
-    return prometheusNamespaces().insert(std::string(prometheus_namespace)).second;
+  if (!std::regex_match(prometheus_namespace.begin(), prometheus_namespace.end(),
+                        namespaceRegex())) {
+    return false;
   }
-  return false;
+  // These names are used as prefixes of native Envoy metrics,
+  // so we must avoid registering them as prometheus namespaces.
+  // Otherwise, the native Envoy metrics end up exposed in Prometheus endpoint
+  // without having "envoy_" prefix.
+  // TODO(mathetake): check the completeness of the list in tests.
+  static const absl::flat_hash_set<std::string> reserved_names = {
+      "auth",        "dns",       "cluster_manager",
```

Comment:
this seems hard to keep up-to-date.

Also our style guide prohibits static non-pods. Any statics should be constexpr. This avoids init-order fiascos.

You can lazy-init structures like this or init them as a manually sorted array of char* or absl::string_view, which you can then use with std::binary_search.

But the larger problem is that this list will be hard to keep up-to-date, particularly with other custom extensions.